File: src/ros2_core/tf2_pkg/tf2_pkg/tf2_data_transformer.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.logging import get_logger
from rclpy.time import Time

from geometry_msgs.msg import PoseStamped
from visualization_msgs.msg import Marker
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener
from tf2_ros import TransformException  # type: ignore

# Required import: Registers the math operations for geometry_msgs
# to allow the Buffer to transform PoseStamped messages.
import tf2_geometry_msgs  # type: ignore

class TF2DataTransformerNode(Node):
    """
    This node simulates a sensor detecting an object. It takes the coordinates
    of the object measured from 'my_dynamic_frame' and transforms them to be 
    relative to the 'world' frame.
    """
    def __init__(self):
        super().__init__('tf2_data_transformer_node')
        self.get_logger().info(f"{self.get_name()} has been started!")

        # 1. Create the Buffer and Listener (Pass the node to track use_sim_time)
        self.tf_buffer = Buffer(node=self)
        self.tf_listener = TransformListener(self.tf_buffer, self)
        
        # Create a publisher to visualize the detected object in RViz
        self.marker_pub = self.create_publisher(Marker, 'detected_object', 10)

        # 2. Create a timer to perform the transformation at 10 Hz (every 0.1 seconds)
        self.timer = self.create_timer(0.1, self.timer_callback)

    def timer_callback(self):
        # --- Simulating Sensor Data ---
        # Simulating an object detected exactly 1 meter in front of the camera.
        sensor_data = PoseStamped()
        
        # Specify the coordinate frame the data was measured in.
        sensor_data.header.frame_id = 'my_dynamic_frame'
        
        # Use Time().to_msg() (equivalent to Time 0) to request the most recent
        # available transform, avoiding extrapolation errors in a single-threaded node.
        sensor_data.header.stamp = Time().to_msg() 
        
        # The coordinates of the object relative to the sensor frame.
        sensor_data.pose.position.x = 1.0
        sensor_data.pose.position.y = 0.0
        sensor_data.pose.position.z = 0.0
        sensor_data.pose.orientation.w = 1.0

        try:
            # --- Core Transformation ---
            # Transform the sensor_data into the global 'world' frame.
            world_pose = self.tf_buffer.transform(
                sensor_data, 
                'world'
            )

            # Extract and print the new coordinates
            x = world_pose.pose.position.x
            y = world_pose.pose.position.y
            z = world_pose.pose.position.z
            
            self.get_logger().info(f"Object in World Frame -> X: {x:.2f}, Y: {y:.2f}, Z: {z:.2f}")

            # --- RViz Visualization ---
            # Publish the marker using the transformed global coordinates.
            self.publish_marker(world_pose.pose)

        except TransformException as ex:
            self.get_logger().warn(f"Could not transform data: {ex}")

    def publish_marker(self, pose):
        """Helper function to create and publish the visualization marker."""
        marker = Marker()
        marker.header.frame_id = 'world'
        marker.header.stamp = self.get_clock().now().to_msg()
        marker.ns = 'sensor_data'
        marker.id = 0
        marker.type = Marker.SPHERE
        marker.action = Marker.ADD
        marker.pose = pose
        
        # Set the marker's scale (0.2 meters wide) and color (red).
        marker.scale.x = 0.2
        marker.scale.y = 0.2
        marker.scale.z = 0.2
        marker.color.r = 1.0
        marker.color.a = 1.0  # Alpha (transparency) must be 1.0 to be visible
        
        self.marker_pub.publish(marker)

def main(args=None) -> None:
    log = get_logger("System")
    node_instance = None

    try:
        log.info("Initializing the ROS2 Client...")
        rclpy.init(args=args)

        log.info("Starting a ROS2 Node...")
        node_instance = TF2DataTransformerNode()
        
        rclpy.spin(node_instance)

    except KeyboardInterrupt:
        log.warn("[CTRL+C]>>> Interrupted by the User.")

    except Exception as e:
        log.error(f"Critical Error: {e}")
    
    finally:
        if node_instance is not None:
            log.info("Destroying the ROS2 Node...")
            node_instance.destroy_node()
            node_instance = None

        if rclpy.ok():
            log.info("Manually shutting down the ROS2 Client...")
            rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/ros2_core/tf2_pkg/src/tf2_data_transformer.cpp
#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "visualization_msgs/msg/marker.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"

// Registers the math operations to allow the Buffer to transform PoseStamped messages.
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"

using namespace std::chrono_literals;

class TF2DataTransformerNode : public rclcpp::Node
{
public:
    TF2DataTransformerNode() : Node("tf2_data_transformer_node")
    {
        RCLCPP_INFO(this->get_logger(), "%s has been started!", this->get_name());

        // 1. Create the Buffer and Listener (Pass the clock for use_sim_time compatibility)
        tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

        // 2. Create the Publisher for the RViz marker
        marker_pub_ = this->create_publisher<visualization_msgs::msg::Marker>("detected_object", 10);
        
        // 3. Create a timer at 10 Hz (every 100ms)
        timer_ = this->create_timer(
            100ms,
            [this]() -> void { this->timer_callback(); }
        );
    }

private:
    void timer_callback()
    {
        // --- Simulating Sensor Data ---
        auto sensor_data = std::make_shared<geometry_msgs::msg::PoseStamped>();
        sensor_data->header.frame_id = "my_dynamic_frame";
        // Using Time(0) to request the most recent available transform
        sensor_data->header.stamp = rclcpp::Time(0); 
        
        sensor_data->pose.position.x = 1.0;
        sensor_data->pose.position.y = 0.0;
        sensor_data->pose.position.z = 0.0;
        sensor_data->pose.orientation.w = 1.0;

        try
        {
            // --- Core Transformation ---
            // Ask the buffer to transform our data into the 'world' frame
            auto world_pose = std::make_shared<geometry_msgs::msg::PoseStamped>();
            *world_pose = tf_buffer_->transform(*sensor_data, "world");

            RCLCPP_INFO(this->get_logger(), "Object in World Frame -> X: %.2f, Y: %.2f, Z: %.2f",
                        world_pose->pose.position.x,
                        world_pose->pose.position.y,
                        world_pose->pose.position.z);

            // --- RViz Visualization ---
            publish_marker(world_pose->pose);
        }
        catch (const tf2::TransformException & ex)
        {
            RCLCPP_WARN(this->get_logger(), "Could not transform data: %s", ex.what());
        }
    }

    void publish_marker(const geometry_msgs::msg::Pose & pose)
    {
        auto marker = std::make_shared<visualization_msgs::msg::Marker>();
        marker->header.frame_id = "world";
        marker->header.stamp = this->get_clock()->now();
        marker->ns = "sensor_data";
        marker->id = 0;
        marker->type = visualization_msgs::msg::Marker::SPHERE;
        marker->action = visualization_msgs::msg::Marker::ADD;
        marker->pose = pose;
        
        marker->scale.x = 0.2;
        marker->scale.y = 0.2;
        marker->scale.z = 0.2;
        marker->color.r = 1.0f;
        marker->color.a = 1.0f;
        
        marker_pub_->publish(*marker);
    }

    std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
    rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr marker_pub_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
    auto log = rclcpp::get_logger("System");
    TF2DataTransformerNode::SharedPtr node_instance = nullptr;

    try {
        RCLCPP_INFO(log, "Initializing the ROS2 Client...");
        rclcpp::init(argc, argv);

        RCLCPP_INFO(log, "Starting a ROS2 Node...");
        node_instance = std::make_shared<TF2DataTransformerNode>();
        rclcpp::spin(node_instance);

        RCLCPP_WARN(log, "[CTRL+C]>>> Interrupted by the User.");
        RCLCPP_INFO(log, "Destroying the ROS2 Node...");
    }
    catch (const std::exception & e) {
        RCLCPP_ERROR(log, "Critical Error: %s", e.what());
    }

    if (rclcpp::ok()) {
        RCLCPP_INFO(log, "Manually shutting down the ROS2 Client...");
        rclcpp::shutdown();
    }
    return 0;
}